fs/operations/operations.go

```diff
@@ -1350,6 +1350,87 @@ func Rmdirs(ctx context.Context, f fs.Fs, dir string, leaveRoot bool) error {
 	return nil
 }
 
+// GetCompareDest sets up --compare-dest
+func GetCompareDest() (CompareDest fs.Fs, err error) {
+	CompareDest, err = cache.Get(fs.Config.CompareDest)
+	if err != nil {
+		return nil, fserrors.FatalError(errors.Errorf("Failed to make fs for --compare-dest %q: %v", fs.Config.CompareDest, err))
+	}
+	return CompareDest, nil
+}
+
+// CompareDest checks --compare-dest see if src needs to
+// be copied
+//
+// Returns True if src is in --compare-dest
+func CompareDest(ctx context.Context, dst, src fs.Object, CompareDest fs.Fs) bool {
+	var remote string
+	if dst == nil {
+		remote = src.Remote()
+	} else {
+		remote = dst.Remote()
+	}
+	CompareDestFile, _ := CompareDest.NewObject(ctx, remote)
+	if Equal(ctx, src, CompareDestFile) {
+		fs.Debugf(src, "Destination found in --compare-dest, skipping")
+		return true
+	}
+	return false
+}
+
+// GetCopyDest sets up --copy-dest
+func GetCopyDest(fdst fs.Fs) (CopyDest fs.Fs, err error) {
+	CopyDest, err = cache.Get(fs.Config.CopyDest)
+	if err != nil {
+		return nil, fserrors.FatalError(errors.Errorf("Failed to make fs for --copy-dest %q: %v", fs.Config.CopyDest, err))
+	}
+	if !SameConfig(fdst, CopyDest) {
+		return nil, fserrors.FatalError(errors.New("parameter to --copy-dest has to be on the same remote as destination"))
+	}
+	if !CanServerSideMove(CopyDest) {
+		return nil, fserrors.FatalError(errors.New("can't use --copy-dest on a remote which doesn't support server side move or copy"))
+	}
+	return CopyDest, nil
+}
+
+// CopyDest checks --copy-dest to see if src needs to
+// be copied
+//
+// Returns True if src was copied from --copy-dest
+func CopyDest(ctx context.Context, fdst fs.Fs, dst, src fs.Object, CopyDest fs.Fs) bool {
+	var remote string
+	if dst == nil {
+		remote = src.Remote()
+	} else {
+		remote = dst.Remote()
+	}
+	CopyDestFile, _ := CopyDest.NewObject(ctx, remote)
+	if equal(ctx, src, CopyDestFile, fs.Config.SizeOnly, fs.Config.CheckSum, false) {
+		if dst == nil || !Equal(ctx, src, dst) {
+			// FIXME BackupDir depends on #3272
+			//if dst != nil && Backupdir != nil {
+			//	err = MoveBackupDir(ctx, backupDir, dst)
+			//	if err != nil {
+			//		return errors.Wrap(err, "moving to --backup-dir failed")
+			//	}
+			//	// If successful zero out the dstObj as it is no longer there
+			//	dst = nil
+			//}
+			_, err := Copy(ctx, fdst, dst, remote, CopyDestFile)
+			if err != nil {
+				fs.Debugf(src, "Destination found in --copy-dest, error copying")
```

Comment:
This should be an `Errorf` log.

Not bubbling up the error means that the sync won't fail and won't be retries if this fails.

I think that the sync should fail if the copy doesn't work - the user explicitly asked for it after all.

So I think this function needs to return an `error`

Reply:
good point